### PR TITLE
doctests in TestCase classes

### PR DESCRIPTION
**Pull request by jimfulton**
Added support for creating doctests as methods of
unittest.TestCase classes so that they can found automatically
by test runners, like *nose* that ignore test suites.

### CODE REVIEW

File: src/zope/testing/doctestcase.py [hunk 1]
```diff
@@ -0,0 +1,137 @@
+r"""Doctests in TestCase classes
+
+The original doctest unittest integration was based on unittest test
+suites, which have falled out of favor. This module provides a way to
```
Comment (mgedmin) [inline src/zope/testing/doctestcase.py]: I've always seen it spelled "have fallen out of favor".  Typo?

File: src/zope/testing/doctestcase.py [hunk 2]
```diff
@@ -0,0 +1,137 @@
...
+  attributes defined in the test case.
+
+- The body of a function who's docstring defines a doctest is run before
+  the test, providing test-specific setup, if desired.
+
+- A the return value of the doctest function can be used to decorate a
+  function.  The decorated function is run before the test is executed.
+  This provides a way to provide some test-specfic setup, if desired.
+
+  Note that the docstring, of any, of the decorated function is ignored.
```
Comment (mgedmin) [inline src/zope/testing/doctestcase.py]: Typo: "if any".

File: src/zope/testing/doctestcase.py [hunk 3]
```diff
@@ -0,0 +1,137 @@
...
+        self.checker = checker
+        self.setups = []
+
+        if isinstance(test, str):
+            if '\n' not in test:
+                base = os.path.dirname(os.path.abspath(
+                    sys._getframe(2).f_globals['__file__']
+                    ))
+                path = os.path.join(base, test)
+                self.test = open(path).read()
```
Comment (mgedmin) [inline src/zope/testing/doctestcase.py]: `with open(path) as fp: self.test = fp.read()`?

File: src/zope/testing/doctestcase.txt [hunk 4]
```diff
@@ -0,0 +1,192 @@
...
+    ...     ''', checker=EasyChecker())
+    ...
+    ...     @DocTest(optionflags=doctest.ELLIPSIS)
+    ...     def test3(self):
+    ...         '''
+    ...         >>> 'Hello'
+    ...         '...'
+    ...         '''
+    ...
+    ...     @DocTest('test4e.txt', optionflags = doctest.ELLIPSIS)
```
Comment (mgedmin) [inline src/zope/testing/doctestcase.txt]: No spaces around `=`.

File: src/zope/testing/tests.py [hunk 5]
```diff
@@ -15,7 +15,7 @@
 import sys
 import re
 import unittest
-from zope.testing import renormalizing
+from zope.testing import renormalizing, doctestcase
```
Comment (mgedmin) [inline src/zope/testing/tests.py]: I don't see where this import is used.

File: src/zope/testing/doctestcase.py [hunk 6]
```diff
@@ -0,0 +1,137 @@
...
+  the test, providing test-specific setup, if desired.
+
+- A the return value of the doctest function can be used to decorate a
+  function.  The decorated function is run before the test is executed.
+  This provides a way to provide some test-specfic setup, if desired.
+
+  Note that the docstring, of any, of the decorated function is ignored.
+
+Also note that, unlike regular unit tests, module globals from the
+module defining the tests are't included in the test globals.
```
Comment (mgedmin) [inline src/zope/testing/doctestcase.py]: Typo: aren't.

(Not sharing globals makes me a bit sad: this leads to much repetition of import statements.)

**Comment by mgedmin** [comment]
nose has `--with-doctest` and `--doctest-tests`.  Granted, that has some disadvantages (e.g. you cannot specify custom checkers or optionflags), so there's room for other solutions.

I have two reservations for this PR:

- the camel-casing of the `@DocText` decorator seems weird to me

- ignoring the module globals for doctests defined in docstrings feels alien (and makes it harder to share import statements and stubs defined in the test module)

**Comment by jimfulton** [comment]
The biggest problems I have wrt current nose support is the poor support for tst-specific test fixtures, which I find unusable.

WRT naming, I'm open to other suggestions.  I was originally going to use doctest, but then that would have been awkward when importing the standard doctest module to get at option flags and othe rmodule contents. doc_test would look weird too.  At least camel-case make sense because this is a constructor.

WRT module globals, I have mixed feelings about this. Would you expose module globals for test files too? I assume not.

A bigger question I had was whether to include test-case instance dictionary data in the doctest globals.  Eventually decided it would be better to expose a ``test`` object, which would give access to test case methods.

**Comment by jimfulton** [comment]
Oh, and another missing feature is the ability to declare execution of doctests from a module.

I'll add that in a later PR.  I'm kinda anxious to get the PR done cuz U need it for work. :)

File: src/zope/testing/doctestcase.py (hunk 1 above)
Comment (jimfulton) [inline src/zope/testing/doctestcase.py]: :) typo fixed

File: src/zope/testing/doctestcase.py (hunk 2 above)
Comment (jimfulton) [inline src/zope/testing/doctestcase.py]: fixed

File: src/zope/testing/doctestcase.py (hunk 3 above)
Comment (jimfulton) [inline src/zope/testing/doctestcase.py]: thanks -- fixed

File: src/zope/testing/doctestcase.txt (hunk 4 above)
Comment (jimfulton) [inline src/zope/testing/doctestcase.txt]: fixed

File: src/zope/testing/tests.py (hunk 5 above)
Comment (jimfulton) [inline src/zope/testing/tests.py]: fixed

File: src/zope/testing/doctestcase.py (hunk 6 above)
Comment (jimfulton) [inline src/zope/testing/doctestcase.py]: fixed

**Comment by jimfulton** [comment]
wrt naming: doctestcase?

**Comment by jimfulton** [comment]
I think having access to module globals makes more sense for traditional doctests that exist in non-test modules.

I'm inclined, for now, to not include module globals, which are somewhat less likely to be interesting in test modules.

It's interesting to note that nose bothers not to look in test modules for doctests, by default.

**Comment by mgedmin** [comment]
As for naming, yeah, the most sense-making choice (`@doctest`) conflicts with using `doctest.ELLIPSIS` etc.  Good point there.

I can get used to `@DocTest`.  I've been using `@Lazy` for a long time without complaining.

As for module globals: right, sharing globals with separate files feels Wrong, just like not letting doctests in docstring access globally defined names feels Wrong to me.   I'd prefer consistency with the stdlib doctest module.

I'm tempted to ask for an option `@DocTest(globals=True)`, but that's a bad way to design APIs, so I won't ;).

**Comment by leorochael** [comment]
I mildly dislike the DWIMing of the first argument acting as either a filename or the test itself, but I can't convince myself that using differently named decorators or different named parameters would make it any better, so I won't complain.

WRT globals, I do think that sharing globals into inlined doctests is important (both the 1st argument tests and the decorated method docstrings), mainly because if I'm mixing doctests with regular test methods in the same test file I'd find it annoying to have to reimport modules that were imported globally only inside the inlined doctests and not in the test methods.

The external doctests I agree shouldn't share globals with the test module.

(I guess a case could be made here that the "test" global should be called (or aliased) "self" in the inlined tests)

WRT naming, `@doctestcase` seems wrong since the decorator creates a test (method), not a test-case (which is a class with several tests, a setup and a teardown). I'd also prefer a lowercased name, but anything I can think of would be weirder or less convenient than `@DocTest`.

All in all, I really like this PR for allowing doctests being able to reuse TestCase setup/teardown and being locatable by `nose`!

**Comment by jimfulton** [comment]
After reading Leonardo's comments, separate function names is growing on me:

  docteststring
  doctestfile
  doctestmethod

doctestmethod and docteststring would inject module globals.

**Comment by jimfulton** [comment]
I prefer test for referring to the test.  Having 'self' at the top level of doctest code seems pretty weird.

Also, if a test defines a class, the self in the example class would hide access to the test.

But perhaps this could work differently for doctestfile vs docteststring and doctestmethod.

File: src/zope/testing/doctestcase.py [hunk 7]
```diff
@@ -0,0 +1,138 @@
...
+  attributes defined in the test case.
+
+- The body of a function who's docstring defines a doctest is run before
+  the test, providing test-specific setup, if desired.
+
+- A the return value of the doctest function can be used to decorate a
+  function.  The decorated function is run before the test is executed.
+  This provides a way to provide some test-specfic setup, if desired.
+
+  Note that the docstring, if any, of the decorated function is ignored.
```
Comment (leorochael) [inline src/zope/testing/doctestcase.py]: The docstring is obviously not ignored in the "naked decorator" case. So this sentence, in isolation, is slightly confusing.

**Comment by leorochael** [comment]
In my mind, `self` in the doctest doesn't look so weird when it's, for instance, the docstring of the test method where the developer just did `self.something = something_else` to prepare for the test. But I agree it only makes sense in the `docteststring/method` cases.

And the hiding of `self` would be undesirable indeed, so if it's there it should be an alias to `test`.

Again, this is minor IMO, and maybe not worth the extra effort.

But speaking of globals, it would be extra nice (perhaps in another pull request) if there was a way of passing new top-level names into the doctest from the decorated function.

**Comment by jimfulton** [comment]
I've rewritten this based on comments:

- Now have separate constructors, doctestmethod, docteststring, and doctestfile, with aliases method,
  string, and file.

- doctestmethods and docteststrings have access to module globals and access the test case 
  instance as self.

- If a test case defines a globs attribute, it must be a dictionary and it's contents are added to test globs.

**Comment by jimfulton** [comment]
BTW, I plan to submit a Python PR after we've used this in zope.testing for a while.

**Comment by jimfulton** [comment]
What do y'all think? :)

File: src/zope/testing/doctestcase.py [hunk 8]
```diff
@@ -0,0 +1,222 @@
...
+
+        @doctestcase.doctestmethod(optionflags=doctest.ELLIPSIS)
+        def test3(self):
+            '''
+            >>> self.a, self.x, g, c
+            (1, 3, 'global', 9)
+            '''
+            self.x = 3
+
+        @doctestcase.doctestfile('test4.txt')
```
Comment (mgedmin) [inline src/zope/testing/doctestcase.py]: Do we really need two different decorators `@doctestcase.file` and `@doctestcase.doctestfile`?

Comment (mgedmin) [inline src/zope/testing/doctestcase.py]: Oh, one's an alias for the other.  Never mind.

File: src/zope/testing/doctestcase.py [hunk 9]
```diff
@@ -0,0 +1,222 @@
...
+docteststring (alias string)
+  docteststring constructs a doctest from a string.
+
+doctestmethod (alias method)
+  doctestmethod constructs a doctest from a method.
+
+  The method's docstring provides the test. The method's body provides
+  optional test-specific setup.
+
+Note that short aliases are provided, which maye be useful in certain
```
Comment (mgedmin) [inline src/zope/testing/doctestcase.py]: Typo: "maye" -> "may"

File: src/zope/testing/doctestcase.py [hunk 10]
```diff
@@ -0,0 +1,222 @@
...
+
+- In tests created with the ``docteststring`` and ``doctestmethod``
+  constructors, the test case instance is available as the ``self``
+  variable.
+
+- In tests created with the ``doctestfile`` constructor, the test case
+  instance is available as the ``test`` variable.
+
+- If a test case defines a globs attribute, it must be a dictionary
+  and it's contents are added to the test globals.
```
Comment (mgedmin) [inline src/zope/testing/doctestcase.py]: Typo: it's -> its.

File: src/zope/testing/doctestcase.py [hunk 11]
```diff
@@ -0,0 +1,222 @@
...
+import types
+
+__all__ = ['doctestmethod', 'docteststring', 'doctestfile']
+
+_parser = doctest.DocTestParser()
+
+def doctestmethod(test=None, optionflags=0, checker=None):
+    """Define a doctest from a method within a unittest.TestCase
+
+    The method's doc string provides the test source. It's body is
```
Comment (mgedmin) [inline src/zope/testing/doctestcase.py]: It's -> Its

File: src/zope/testing/doctestcase.py [hunk 12]
```diff
@@ -0,0 +1,222 @@
...
+    called before the test and may perform test-specific setup.
+
+    You can pass doctest option flags and a custon checker.
+
+    Variables defined in the enclosing module are available in the test.
+
+    If a test case defines a globs attribute, it must be a dictionary
+    and it's contents are added to the test globals.
```
Comment (mgedmin) [inline src/zope/testing/doctestcase.py]: it's -> its

File: src/zope/testing/doctestcase.py [hunk 13]
```diff
@@ -0,0 +1,222 @@
...
+
+def docteststring(test, optionflags=0, checker=None):
+    """Define a doctest from a string within a unittest.TestCase
+
+    You can pass doctest option flags and a custon checker.
+
+    Variables defined in the enclosing module are available in the test.
+
+    If a test case defines a globs attribute, it must be a dictionary
+    and it's contents are added to the test globals.
```
Comment (mgedmin) [inline src/zope/testing/doctestcase.py]: it's -> its

File: src/zope/testing/doctestcase.py [hunk 14]
```diff
@@ -0,0 +1,222 @@
...
+
+    return test_string
+
+string = docteststring
+
+def doctestfile(path, optionflags=0, checker=None):
+    """Define a doctest from a test file within a unittest.TestCase
+
+    The file path may be relative or absolute. If it's relative (the
+    common case), it will interpreted relative to the directory
```
Comment (mgedmin) [inline src/zope/testing/doctestcase.py]: will -> will be

File: src/zope/testing/doctestcase.py [hunk 15]
```diff
@@ -0,0 +1,222 @@
...
+    containing the referencing module.
+
+    You can pass doctest option flags and a custon checker.
+
+    If a test case defines a globs attribute, it must be a dictionary
+    and it's contents are added to the test globals.
```
Comment (mgedmin) [inline src/zope/testing/doctestcase.py]: it's -> its

File: src/zope/testing/doctestcase.py [hunk 16]
```diff
@@ -0,0 +1,222 @@
...
+
+def _run_test(self, test, globs, name, path,
+              optionflags, checker, testname='self', lineno=0):
+    globs.update(getattr(self, 'globs', ()))
+    globs[testname] = self
+    optionflags |= doctest.IGNORE_EXCEPTION_DETAIL
+    doctest.DocTestCase(
+        _parser.get_doctest(test, globs, name, path, lineno),
+        optionflags = optionflags,
+        checker = checker,
```
Comment (mgedmin) [inline src/zope/testing/doctestcase.py]: PEP-8: no spaces around `=` in keyword args

File: src/zope/testing/doctestcase.txt [hunk 17]
```diff
@@ -0,0 +1,219 @@
...
+
+- In tests created with the ``docteststring`` and ``doctestmethod``
+  constructors, the test case instance is available as the ``self``
+  variable.
+
+- In tests created with the ``doctestfile`` constructor, the test case
+  instance is available as the ``test`` variable.
+
+- If a test case defines a globs attribute, it must be a dictionary
+  and it's contents are added to the test globals.
```
Comment (mgedmin) [inline src/zope/testing/doctestcase.txt]: it's -> its

File: src/zope/testing/doctestcase.txt [hunk 18]
```diff
@@ -0,0 +1,219 @@
...
+
+The constructors accept standard doctest ``optionflags`` and
+``checker`` arguments.
```
Comment (mgedmin) [inline src/zope/testing/doctestcase.txt]: Might be worth mentioning somewhere that IGNORE_EXCEPTION_DETAIL is always implicitly OR'ed in.

**Comment by mgedmin** [comment]
Looks good to me (other than some typos)!

I didn't mention that some docstings ought to have a period at the end of the first sentence (as per PEP-257), since I didn't want to appear too nitpickery.

File: src/zope/testing/doctestcase.txt (hunk 18 above)
Comment (jimfulton) [inline src/zope/testing/doctestcase.txt]: Added a note.

File: src/zope/testing/doctestcase.py (hunk 10 above)
Comment (jimfulton) [inline src/zope/testing/doctestcase.py]: This and other typos fixed.

File: src/zope/testing/doctestcase.py (hunk 16 above)
Comment (jimfulton) [inline src/zope/testing/doctestcase.py]: I prefer spaces around =s when options are on their own lines, but changed.

**Comment by jimfulton** [comment]
Thanks for the comments! I've addresses all of them.

**Comment by leorochael** [comment]
Didn't do as thorough a review as @mgedmin, but looks nice.

The independent functions are nicer and more readable than before.

**Pull request merged**